examples/sparse/bal_utils.py

```diff
@@ -0,0 +1,160 @@
+"""
+This file contains the helper functions for the Bundle Adjustment in the Large dataset.
+
+The dataset is from the following paper:
+Sameer Agarwal, Noah Snavely, Steven M. Seitz, and Richard Szeliski.
+Bundle adjustment in the large.
+In European Conference on Computer Vision (ECCV), 2010.
+
+Link to the dataset: https://grail.cs.washington.edu/projects/bal/
+"""
+
+import os, torch
+import pypose as pp
+import matplotlib.pyplot as plt
+from bal_loader import build_pipeline
+
+def reprojerr(pose, points, pixels, intrinsics, distortions, point_index, camera_index):
+    """
+    Calculates batched per-pixel reprojection error considering camera distortion.
+    Using BAL's suggestion:
+
+    We use a pinhole camera model; the parameters we estimate for each camera area rotation R, a translation t, a focal length f and two radial distortion parameters k1 and k2. The formula for projecting a 3D point X into a camera R,t,f,k1,k2 is:
+    P  =  R * X + t       (conversion from world to camera coordinates)
+    p  = -P / P.z         (perspective division)
+    p' =  f * r(p) * p    (conversion to pixel coordinates)
+    where P.z is the third (z) coordinate of P. In the last equation, r(p) is a function that computes a scaling factor to undo the radial distortion:
+    r(p) = 1.0 + k1 * ||p||^2 + k2 * ||p||^4.
+
+    Args:
+        pose (``LieTensor``): The camera extrinsics.
+            The shape has to be (..., 7).
+        points (``torch.Tensor``): The object points in world coordinate.
+            The shape has to be (..., N, 3).
+        pixels (``torch.Tensor``): The image points. The associated pixel.
+            The shape has to be (..., N, 2).
+        intrinsics (``torch.Tensor``): intrinsic matrices.
+            The shape has to be (..., 3, 3).
+        distortions (``torch.Tensor``): The camera distortions.
+            The shape has to be (..., 2).
+        point_index (``torch.Tensor``): The indices of points.
+            The shape has to be (..., N).
+        camera_index (``torch.Tensor``): The indices of cameras.
+            The shape has to be (..., N).
+    Returns:
+        Per-pixel reprojection error. The shape is (..., N).
+    """
+    # convert to camera coordinates
```

Comment:
Let's have a quick meeting to discuss how to do proper comments to this function.

I have a hard time understanding the magic dimension numbers used in this function. I think future users might try to take a look at how projection errors are implemented and they might run into the same problem.

Reply:
After a discussion, Yaoyu and I reached the conclusion that we should integrate the image undistortion function into pypose's geometry module, and then simply call pypose native functions here. This will avoid exposing the complexity of raw dimension manipulations to the users and also strengthen pypose's API library.

A refactor is on the way.